Library/OcAppleImg4Lib/OcAppleImg4Lib.c

```diff
@@ -0,0 +1,316 @@
+  @param[out] DigestSize      On output, the size, in bytes, of *HashDigest.
+
+  @retval EFI_SUCCESS             ImageBuffer is correctly signed.
+  @retval EFI_INVALID_PARAMETER   One or more required parameters are NULL.
+  @retval EFI_OUT_OF_RESOURCES    Not enough resources are available.
+  @retval EFI_SECURITY_VIOLATION  ImageBuffer's signature is invalid.
+
+**/
+EFI_STATUS
+EFIAPI
+AppleImg4Verify (
+  IN  APPLE_IMG4_VERIFICATION_PROTOCOL  *This,
+  IN  UINT32                            ObjType,
+  IN  CONST VOID                        *ImageBuffer,
+  IN  UINTN                             ImageSize,
+  IN  UINT8                             SbMode,
+  IN  CONST VOID                        *ManifestBuffer,
+  IN  UINTN                             ManifestSize,
+  OUT UINT8                             **HashDigest OPTIONAL,
+  OUT UINTN                             *DigestSize OPTIONAL
+  )
+{
+  DERReturn          DerResult;
+
+  DERImg4Environment  EnvInfo;
+  DERImg4ManifestInfo ManInfo;
+  UINT8              ImageDigest[OC_MAX_SHA_DIGEST_SIZE];
+
+  if ((ImageBuffer    == NULL || ImageSize    == 0)
+   || (ManifestBuffer == NULL || ManifestSize == 0)
+   || (HashDigest     == NULL && DigestSize   != NULL)
+   || (HashDigest     != NULL && DigestSize   == NULL)
+   || (SbMode         != AppleImg4SbModeMedium
+    && SbMode         != AppleImg4SbModeFull)) {
+    return EFI_INVALID_PARAMETER;
+  }
+
+  DerResult = DERImg4ParseManifest (
+                &ManInfo,
+                ManifestBuffer,
+                ManifestSize,
+                ObjType
+                );
+  if (DerResult != DR_Success) {
+    return EFI_SECURITY_VIOLATION;
+  }
+
+  if (ManInfo.imageDigestSize == SHA512_DIGEST_SIZE) {
+    Sha512 (ImageDigest, ImageBuffer, ImageSize);
+  } else if (ManInfo.imageDigestSize == SHA384_DIGEST_SIZE) {
+    Sha384 (ImageDigest, ImageBuffer, ImageSize);
+  } else if (ManInfo.imageDigestSize == SHA256_DIGEST_SIZE) {
+    Sha256 (ImageDigest, ImageBuffer, ImageSize);
+  } else {
+    return EFI_SECURITY_VIOLATION;
+  }
+
+  InternalRetrieveHwInfo (&EnvInfo);
+
+  if (SbMode == AppleImg4SbModeMedium) {
+    if (ManInfo.hasEcid
+     || !ManInfo.hasXugs || ManInfo.environment.xugs != EnvInfo.xugs
+     || (ManInfo.environment.internalUseOnlyUnit
+      && ManInfo.environment.internalUseOnlyUnit != EnvInfo.internalUseOnlyUnit)) {
+      return EFI_SECURITY_VIOLATION;
+    }
+  } else if (SbMode == AppleImg4SbModeFull) {
+    if (!ManInfo.hasEcid || ManInfo.environment.ecid != EnvInfo.ecid
+     || ManInfo.hasXugs
+     || ManInfo.environment.internalUseOnlyUnit) {
+      return EFI_SECURITY_VIOLATION;
+    }
+  }
+
+  if (ManInfo.environment.boardId                != EnvInfo.boardId
+   || ManInfo.environment.chipId                 != EnvInfo.chipId
+   || ManInfo.environment.certificateEpoch        < EnvInfo.certificateEpoch
+   || ManInfo.environment.productionStatus        < EnvInfo.productionStatus
```

Comment:
That's a boolen, why <? Same with securityMode. Please review that code with the RE and valid value range. Perhaps it is either not a boolean but UINT8 with 0, 1, [future extension] range, or something is wrong here.

Reply:
Interesting, I haven't even noticed. I just verified my code complies to the RE. I guess it could be related to IMG4 Booleans using 0 and 0xFF? Maybe we should have a dedicated IMG4 bool type?

Reply:
Sorry, I was in a rush when I wrote that, the first part makes little sense. TRUE is the secure default, so if the Manifest allows a relaxation, the stricter setting is still tolerated, which is why "<" is used. As for the dedicated type, it might make sense considering the 0 / 0xFF encoding, however if we treat bools by their evalutation in expressions (0 vs != 0) rather than expression results (only ever 0 or 1), this is still fine.

Reply:
We must use dedicated types for this encoding, I think it may actually be the case they are not bools.

Reply:
Agreed

Reply:
Yes, I believe it must not be BOOLEAN, regardless of whatever sources we have.

Reply:
Why must it not be boolean if the linked code actually returns a boolean? The 0 / 0xFF encoding point was my bad, this never leaves libDER internals.